include/sys/onoff.h

```diff
@@ -0,0 +1,606 @@
+ * operation succeeds the result of the reset operation itself is
+ * provided through the configured client notification method,
+ * possibly before this call returns.  Multiple clients may request a
+ * reset; all are notified when it is complete.
+ *
+ * Note that the call to this function may succeed in a case where the
+ * actual reset fails.  Always check the operation completion result.
+ *
+ * This function is blocking if the reset transition is blocking,
+ * unless client notification specifies no-wait.
+ *
+ * @note Due to the conditions on state transition all incomplete
+ * asynchronous operations will have been informed of the error when
+ * it occurred.  There need be no concern about dangling requests left
+ * after a reset completes.
+ *
+ * @param srv the service to be reset.
+ *
+ * @param cli pointer to client state, including instructions on how
+ * to notify the client when reset completes.  Behavior is undefined
+ * if cli references an object associated with an incomplete service
+ * operation.
+ *
+ * @retval 0 on success
+ * @retval -ENOTSUP if reset is not supported
+ * @retval -EINVAL if the parameters are invalid, or if the service
+ * does not have a recorded error
+ */
+int onoff_service_reset(struct onoff_service *srv,
+			struct onoff_client *cli);
+
+/**
+ * @brief Attempt to cancel an in-progress client operation.
+ *
+ * It may be that a client has initiated an operation but needs to
+ * shut down before the operation has completed.  For example, when a
+ * request was made and the need is no longer present.
+ *
+ * There is limited support for cancelling an in-progress operation:
+ * * If a start or reset is in progress, all but one clients
+ *   requesting the start can cancel their request.
+ * * If a stop is in progress, all clients requesting a restart can
+ *   cancel their request;
+ * * A client requesting a release cannot cancel the release.
+ *
+ * Be aware that any transition that was initiated on behalf of the
+ * client will continue to progress to completion.  The restricted
+ * support for cancellation ensures that for any in-progress
+ * transition there will always be at least one client that will be
+ * notified when the operation completes.
+ *
+ * If the cancellation fails the service retains control of the client
+ * object, and the client must wait for operation completion.
+ *
+ * @param srv the service for which an operation is to be cancelled.
+ *
+ * @param cli a pointer to the same client state that was provided
+ * when the operation to be cancelled was issued.  If the cancellation
+ * is successful the client will be notified of operation completion
+ * with a result of `-ECANCELED`.
+ *
+ * @retval 0 if the cancellation was completed before the client could
+ * be notified.  The client will be notified through cli with an
+ * operation completion of `-ECANCELED`.
+ * @retval -EINVAL if the parameters are invalid.
+ * @retval -EWOULDBLOCK if cancellation was rejected because the
```

Comment:
The user reaction must be to continue to wait for the result: the operation cannot be cancelled.  

I think the detail is warranted: the last client canceling is the only condition where this would happen: cancels requested while in-progress that are not the last one are processed immediately.